tests/integration/sync/sync_integ_base.py

```diff
@@ -1,19 +1,151 @@
+                self._delete_companion_stack(cfn_client, ecr_client, self._stack_name_to_companion_stack(stack_name))
+                cfn_client.delete_stack(StackName=stack_name)
         super().tearDown()
 
+    def _get_stacks(self, stack_name):
+        try:
+            physical_ids = {}
+            response = self.cfn_client.describe_stack_resources(StackName=stack_name).get("StackResources", {})
+            for resource in response:
+                resource_type = resource.get("ResourceType")
+                if resource_type == "AWS::CloudFormation::Stack":
+                    nested_stack_physical_id = resource.get("PhysicalResourceId")
+                    nested_stack_name = nested_stack_physical_id.split("/")[1]
+                    nested_stack_physical_ids = self._get_stacks(nested_stack_name)
+                    for nested_resource_type, nested_physical_ids in nested_stack_physical_ids.items():
+                        if nested_resource_type not in physical_ids:
+                            physical_ids[nested_resource_type] = []
+                        physical_ids[nested_resource_type] += nested_physical_ids
+                    continue
+                if resource_type not in physical_ids:
+                    physical_ids[resource.get("ResourceType")] = []
+                physical_ids[resource_type].append(resource.get("PhysicalResourceId"))
+            return physical_ids
+        except ClientError:
+            pass
+        return None
+
+    def _get_lambda_response(self, lambda_function):
+        try:
+            lambda_response = self.lambda_client.invoke(FunctionName=lambda_function, InvocationType="RequestResponse")
+            payload = json.loads(lambda_response.get("Payload").read().decode("utf-8"))
+            return payload.get("body")
+        except ClientError as ce:
+            LOG.error(ce)
+        except Exception as e:
+            LOG.error(e)
+        return ""
+
+    def _get_api_message(self, rest_api):
+        try:
+            api_resource = self.api_client.get_resources(restApiId=rest_api)
+            for item in api_resource.get("items"):
+                if "GET" in item.get("resourceMethods", {}):
+                    resource_id = item.get("id")
+                    break
+            self.api_client.flush_stage_cache(restApiId=rest_api, stageName="prod")
+        except ClientError as ce:
+            LOG.error(ce)
+        except Exception as e:
+            LOG.error(e)
+
+        # RestApi deployment needs a wait time before being responsive
+        count = 0
+        while count < 20:
+            try:
+                time.sleep(1)
+                api_response = self.api_client.test_invoke_method(
+                    restApiId=rest_api, resourceId=resource_id, httpMethod="GET"
+                )
+                return api_response.get("body")
+            except ClientError as ce:
+                if count == 20:
+                    LOG.error(ce)
```

Comment:
On 20th failure, can we raise the exception instead of just logging it and returning empty string? The test failure message would be `"" != "expected output"` instead of the exception directly. This will be harder to find the root cause as we need to go back and check the error log.

Reply:
Yep changing it to raise as well. But the thing is in real life customers are not guaranteed to have their rest api updated within 20 seconds. So this might become a flaky test in the future where we see failures randomly